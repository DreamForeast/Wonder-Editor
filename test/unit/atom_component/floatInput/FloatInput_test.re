open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("FloatInput", () => {
    let _triggerChangeInputEvent = (value, domChildren) => {
      let input = WonderCommonlib.ArrayService.unsafeGet(domChildren, 1);
      BaseEventTool.triggerChangeEvent(
        input,
        BaseEventTool.buildFormEvent(value),
      );
    };
    let _triggerBlurEvent = (value, domChildren) => {
      let input = WonderCommonlib.ArrayService.unsafeGet(domChildren, 1);
      BaseEventTool.triggerBlurEvent(
        input,
        BaseEventTool.buildFormEvent(value),
      );
    };
    describe("test FloatInput component set float value", () => {
      test("if float value's decimal digits <= 6, can set the whole value", () => {
        let component =
          ReactTestRenderer.create(
            <FloatInput defaultValue="2" label="xyz" />,
          );
        BaseEventTool.triggerComponentEvent(
          component,
          _triggerChangeInputEvent("351687.54654"),
        );
        BaseEventTool.triggerComponentEvent(
          component,
          _triggerBlurEvent("351687.54654"),
        );
        component |> ReactTestTool.createSnapshotAndMatch;
      });
      test("else, can't set the value", () => {
        let component =
          ReactTestRenderer.create(
            <FloatInput defaultValue="0" label="xyz" />,
          );
        BaseEventTool.triggerComponentEvent(
          component,
          _triggerChangeInputEvent("3.524584654"),
        );
        component |> ReactTestTool.createSnapshotAndMatch;
      });
    });
    describe("deal with the specific case", () => {
      let sandbox = getSandboxDefaultVal();
      beforeEach(() => sandbox := createSandbox());
      afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));
      test("key in value '', shouldn't trigger event", () => {
        let onChange = createEmptyStubWithJsObjSandbox(sandbox);
        let onBlur = createEmptyStubWithJsObjSandbox(sandbox);
        let component =
          ReactTestRenderer.create(
            <FloatInput defaultValue="22" label="xyz" onChange onBlur />,
          );
        BaseEventTool.triggerComponentEvent(
          component,
          _triggerChangeInputEvent(""),
        );
        BaseEventTool.triggerComponentEvent(
          component,
          _triggerBlurEvent(""),
        );
        onChange |> expect |> not_ |> toCalled;
      });
      test("key in value '-', shouldn't trigger event", () => {
        let onChange = createEmptyStubWithJsObjSandbox(sandbox);
        let onBlur = createEmptyStubWithJsObjSandbox(sandbox);
        let component =
          ReactTestRenderer.create(
            <FloatInput defaultValue="22" label="xyz" onChange onBlur />,
          );
        BaseEventTool.triggerComponentEvent(
          component,
          _triggerChangeInputEvent("-"),
        );
        BaseEventTool.triggerComponentEvent(
          component,
          _triggerBlurEvent("-"),
        );
        onChange |> expect |> not_ |> toCalled;
      });
    });
  });