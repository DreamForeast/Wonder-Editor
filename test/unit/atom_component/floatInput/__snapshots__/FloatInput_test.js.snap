// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatInput test FloatInput component set float value else, can't set the value 1`] = `
<article
  className="wonder-float-input"
>
  <span
    className="component-label"
  >
    xyz : 
  </span>
  <input
    className="input-component float-input"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value="0"
  />
</article>
`;

exports[`FloatInput test FloatInput component set float value if float value's decimal digits <= 6, can set the whole value 1`] = `
<article
  className="wonder-float-input"
>
  <span
    className="component-label"
  >
    xyz : 
  </span>
  <input
    className="input-component float-input"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value="351687.54654"
  />
</article>
`;
